docs/snapshot-versioning.md

```diff
@@ -0,0 +1,96 @@
+# Firecracker snapshot versioning
+
+This document describes how Firecracker persists it's state across multiple versions, diving deep into the snapshot format, encoding, compatibility and limitations. 
+
+## Introduction
+The design behind the snapshot implementation enables version tolerant save and restore across multiple Firecracker versions which we call a version space. Basically, one can pause a microvm, save it to disk with Firecracker version **v4** and later load it in Firecracker version **v2**. It also works  in reverse: Firecracker version **v4** loads what  **v2** saves. 
+
+Below is an example graph showing backward and forward snapshot compatibility for a version space formally represented as [v2; vn]. This is the general picture, but keep in mind that future corner cases might remove edges from this graph.
+
+![Version graph](
+images/version_graph.png?raw=true
+"Version graph")
+
+## Overview
+Firecracker persists the microvm state as 2 separate objects:
+ - a  **guest memory** file
+ - a **vmstate** file
+ 
+*The block devices attached to the microvm are not considered part of the state and need to be managed separately.*
+
+### Guest memory
+The guest memory file is agnostic to Firecracker's version tolerant save/restore as it's contents are fully owned by the guest OS. The microVM memory is by choice saved as a dump of each page or as a sparse file of guest modified pages only. 
+### Vmstate
+Stores the internal state of the VMM (device emulation , KVM and vCPUs) with 2 expcetions - serial emulation and vsock backend. 
+
+While we continuously improve and extend Firecracker's features by adding new options, devices or enhancements, the vmstate file changes both structurally and semantically with each new release. The state file includes versioning information and each Firecracker release implements distinct save/load logic for the supported version space. 
```

Comment:
It might not be the case that the vmstate changes with each release.

Reply:
That is correct, but we can still bump the version if there are no changes so we keep the 1:1 promise. The main challenge here is to define a function that maps the Firecracker version string in the API into an interval of versions that match the VersionMap root version.
I think we need to redesign the VersionMap to account for version gaps for that to work.